drivers/qcommsimmodem/radio-settings.c

```diff
@@ -0,0 +1,298 @@
+/*
+ *
+ *  oFono - Open Source Telephony
+ *
+ *  Copyright (C) 2014  Canonical Ltd.
+ *  Copyright (C) 2015 Ratchanan Srirattanamet
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License version 2 as
+ *  published by the Free Software Foundation.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#define _GNU_SOURCE
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+
+#include <glib.h>
+
+#include <ofono/log.h>
+#include <ofono/modem.h>
+#include <ofono/radio-settings.h>
+
+#include "gril.h"
+#include "grilrequest.h"
+#include "grilreply.h"
+
+#include "drivers/rilmodem/radio-settings.h"
+#include "drivers/rilmodem/rilutil.h"
+#include "qcom_msim_modem.h"
+
+struct pending_pref_setting {
+	struct ofono_radio_settings *rs;
+	int pref;
+	int pending_gsm_pref_remaining;
+	struct cb_data *cbd;
+};
+
+struct set_2g_rat {
+	struct ofono_radio_settings *rs;
+	struct pending_pref_setting *pps;
+};
+
+#define MULTISIM_RS_LAST 2
```

Comment:
This should be defined in a header specific to your driver, something like drivers/qcommsimmodem/qcomutil.h, so you can share with other atoms if/when needed.